### PR TITLE
Fix paragraph alignment

### PR DESCRIPTION
**Pull request by smitsk**
Fixes: flutter/flutter#91695 (part 3)

Most likely behavior for left aligned rtl text should be similar to right aligned ltr text. When whitespaces are typed at the end of input it's not shown and text won't moves outside the container.

```
TextField(
  controller: TextEditingController(text: 'input      '),
  decoration: const InputDecoration(filled: true),
  textDirection: TextDirection.ltr,
  textAlign: TextAlign.right,
  maxLines: null,
),
```

I'm going to add a test if this is the correct solution in your opinion.

flutter/flutter#91698 is required for correct work this changes.

### CODE REVIEW

**Comment by justinmc** [comment]
It seems like this doesn't fully fix issue 3 in https://github.com/flutter/flutter/issues/91695.  Below is what I get on native Mac vs. Flutter running this engine code using the example in https://github.com/flutter/flutter/issues/91695#issuecomment-942259114.  It seems like maybe we just need to also not show the whitespace until more text is entered?

| Native Mac TextEdit | Flutter on Mac |
| --- | --- |
| <video src="https://user-images.githubusercontent.com/389558/139729369-606dac9a-d0c0-4d00-82b3-85d3d85773c1.mov" /> | <video src="https://user-images.githubusercontent.com/389558/139729372-0cc9d4c9-633c-434a-9259-3023572c7ac8.mov" /> |



CC @chinmaygarde

**Review by justinmc — COMMENTED** [review]
It seems like this doesn't fully fix issue 3 in https://github.com/flutter/flutter/issues/91695.  Below is what I get on native Mac vs. Flutter running this engine code using the example in https://github.com/flutter/flutter/issues/91695#issuecomment-942259114.  It seems like maybe we just need to also not show the whitespace until more text is entered?

*Edit*: I wonder how https://github.com/flutter/flutter/pull/91698 would affect this behavior.

| Native Mac TextEdit | Flutter on Mac with this PR |
| --- | --- |
| <video src="https://user-images.githubusercontent.com/389558/139729369-606dac9a-d0c0-4d00-82b3-85d3d85773c1.mov" /> | <video src="https://user-images.githubusercontent.com/389558/139729372-0cc9d4c9-633c-434a-9259-3023572c7ac8.mov" /> |


CC @chinmaygarde

**Comment by smitsk** [comment]
> It seems like maybe we just need to also not show the whitespace until more text is entered?
> Edit: I wonder how flutter/flutter#91698 would affect this behavior.

Yes, changes in flutter/flutter#91698 does it. I explained this here https://github.com/flutter/flutter/pull/91698#issuecomment-957955955
I added this requirement in the PR description.

**Review by LongCatIsLooong — COMMENTED** [review]
FYI if you make changes to libtxt here,  https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/modules/skparagraph/ needs to be brought to parity.

File: third_party/txt/src/txt/paragraph_txt.cc [hunk 1]
```diff
@@ -733,7 +733,9 @@ void ParagraphTxt::Layout(double width) {
     size_t line_end_index =
         (paragraph_style_.effective_align() == TextAlign::right ||
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply third_party/txt/src/txt/paragraph_txt.cc:734]: Oh I was wrong UAX#9 says 

> trailing whitespace will appear at the visual end of the line (in the paragraph direction)

**Comment by chinmaygarde** [comment]
@LongCatIsLooong should we wait for the SkParagraph switch or is this good to go?

**Comment by LongCatIsLooong** [comment]
The change makes sense to me (although I think when `paragraph_style_.effective_align()` is `TextAlign::right` and `TextAlign::right` the logic should be symmetric), but I don't know the code very well, and I'm not sure if we should hold off and wait for the SkParagraph switch /cc @jason-simmons @GaryQian  any advice on whether to land this?

**Comment by jason-simmons** [comment]
In general I'd prefer to postpone text related changes until after the switch to SkParagraph is complete

**Comment by chinmaygarde** [comment]
Closing this as [we are not going to accept in libtxt](https://github.com/flutter/engine/pull/29429#issuecomment-966682266).

**Comment by smitsk** [comment]
@jason-simmons @LongCatIsLooong is it good that for web it already works symmetrically without this PR, but for other platforms it is different?

For example:
```
Column(
    mainAxisSize: MainAxisSize.min,
    children: [
      TextField(
        controller: TextEditingController(text: 'input                '),
        decoration: const InputDecoration(filled: true),
        textDirection: TextDirection.ltr,
        textAlign: TextAlign.right,
        maxLines: null,
      ),
      TextField(
        controller: TextEditingController(text: 'input                '),
        decoration: const InputDecoration(filled: true),
        textDirection: TextDirection.rtl,
        textAlign: TextAlign.left,
        maxLines: null,
      ),
    ],
),
```
If running on web (stable channel):

If running on Android (stable channel):


Perhaps this PR could make the behavior more consistent and fix that text is outside the field?

**Comment by LongCatIsLooong** [comment]
As far as I know web uses canvaskit (SkParagraph) by default so if your web result image is generated using the canvaskit backend then when other platforms switch to SkParagraph the bug will probably be gone too.

**Comment by smitsk** [comment]
Well, in this case the PR can really be closed.

**Pull request closed without merging**